aptos-move/dynamic-transaction-composer/src/builder.rs

```diff
@@ -0,0 +1,593 @@
+            .map_err(|err| JsValue::from(format!("{:?}", err)))
+    }
+}
+
+impl TransactionComposer {
+    pub fn insert_module(&mut self, module: CompiledModule) {
+        self.modules.insert(module.self_id(), module);
+    }
+
+    fn check_argument_compatibility(
+        &mut self,
+        argument: &AllocatedLocal,
+        expected_ty: &SignatureToken,
+    ) -> anyhow::Result<()> {
+        let local_ty = if argument.is_parameter {
+            self.parameters_ty[argument.local_idx as usize].clone()
+        } else {
+            self.locals_ty[argument.local_idx as usize].clone()
+        };
+
+        let ty = match argument.op_type {
+            ArgumentOperation::Borrow => SignatureToken::Reference(Box::new(local_ty)),
+            ArgumentOperation::BorrowMut => SignatureToken::MutableReference(Box::new(local_ty)),
+            ArgumentOperation::Copy => {
+                let ability = BinaryIndexedView::Script(self.builder.as_script())
+                    .abilities(&local_ty, &[])
+                    .map_err(|_| anyhow!("Failed to calculate ability for type"))?;
+                if !ability.has_copy() {
+                    bail!("Trying to copy move values that does NOT have copy ability");
+                }
+                local_ty
+            },
+            ArgumentOperation::Move => {
+                if !argument.is_parameter {
+                    if self.locals_availability[argument.local_idx as usize] {
+                        self.locals_availability[argument.local_idx as usize] = false;
+                    } else {
+                        bail!("Trying to use a Move value that has already been moved");
+                    }
+                }
+                local_ty
+            },
+        };
+
+        if &ty != expected_ty {
+            bail!("Type mismatch when passing arugments around");
+        }
+        Ok(())
+    }
+
+    pub fn add_batched_call(
+        &mut self,
+        module: String,
+        function: String,
+        ty_args: Vec<String>,
+        args: Vec<Argument>,
+    ) -> anyhow::Result<Vec<Argument>> {
+        let ty_args = ty_args
+            .iter()
+            .map(|s| TypeTag::from_str(s))
+            .collect::<anyhow::Result<Vec<_>>>()?;
+        let module = ModuleId::from_str(&module)?;
+        let function = Identifier::new(function)?;
+
+        let target_module = match self.modules.get(&module) {
+            Some(module) => module,
+            None => {
+                bail!("Module {} is not yet loaded", module);
+            },
+        };
+
+        let call_idx = self
+            .builder
+            .import_call_by_name(function.as_ident_str(), target_module)?;
+        let call_type_params = ty_args
+            .iter()
+            .map(|ty| import_type_tag(&mut self.builder, ty, &self.modules))
+            .collect::<PartialVMResult<Vec<_>>>()?;
+
+        let subst_mapping = call_type_params
```

Comment:
As I mentioned in a previous comment, would it make more sense to just take in the substitution as `&[SignatureToken]`?

Reply:
I mean, if we have another client that wants to use `BTreeMap` we can create a `trait SubstitutionMap` so that the function supports both data structures

Reply:
Agreed. Changed it to &[SignatureToken]